### PR TITLE
Universal instrument for Java and .NET including .NET6.0

### PR DESCRIPTION
**Pull request by nine5two7**
What and why?
This PR is to support Java and .NET universal instrumentation . In addition, it also tries to support dotnet6.
For new extensions (>= version 24), cases handled include: 1) New trace version, 2) Old trace version, and 3) Undefined trace version
For old extension versions, the previous workflows are still used.

How?
Check the extension version and trace version for Java and .NET runtime
Create a separate versionChecker to verify whether an extension version supports universal instrument, and a trace version is compatible with a given extension version

https://datadoghq.atlassian.net/browse/SLS-2202
https://datadoghq.atlassian.net/browse/SLS-2308

### Review checklist

- [x] Feature or bugfix MUST have appropriate tests (unit, integration)
- [ ] \[_Synthetics_\] New releases of `datadog-ci` MUST be followed by a [synthetics-ci-github-action](https://github.com/DataDog/synthetics-ci-github-action) release

### CODE REVIEW

File: src/commands/lambda/__tests__/instrument.test.ts [hunk 1]
```diff
@@ -1504,50 +1504,6 @@ ${red('[Error]')} Couldn't fetch Lambda functions. Error: Max retry count exceed
 `)
       })
 
-      test('aborts early when a layer version is set for Java', async () => {
```
Comment (nine5two7) [inline src/commands/lambda/__tests__/instrument.test.ts:1507]: Now, a trace layer can be configured for Java.

File: src/commands/lambda/__tests__/functions/instrument.test.ts [hunk 2]
```diff
@@ -458,7 +458,181 @@ describe('instrument', () => {
         '`apiKeySecretArn` is not supported for Node runtimes when using Synchronous Metrics. Use either `apiKey` or `apiKmsKey`.'
       )
     })
+
+    describe('test the new universal instrumentation workflow for Java and .Net', () => {
```
Comment (DarcyRaynerDD) [inline src/commands/lambda/__tests__/functions/instrument.test.ts:462]:
```suggestion
    describe('test universal instrumentation workflow for Java and .Net', () => {
```

File: src/commands/lambda/__tests__/functions/instrument.test.ts [hunk 3]
```diff
@@ -458,7 +458,181 @@ describe('instrument', () => {
...
+        flushMetricsToLogs: false,
+        layerAWSAccount: mockAwsAccount,
+        layerVersion: 13,
+        mergeXrayTraces: false,
+        tracingEnabled: false,
+      }
+      const dotnetRuntime = 'dotnet6'
+      const javaRuntime = 'java11'
+
+      describe('test for runtime ${dotnetRuntime}', () => {
```
Comment (DarcyRaynerDD) [inline src/commands/lambda/__tests__/functions/instrument.test.ts:480]: I think you have to use backticks for template strings to work in javascript
```suggestion
      describe(`test for runtime ${dotnetRuntime}`, () => {
```

File: src/commands/lambda/__tests__/functions/instrument.test.ts [hunk 4]
```diff
@@ -458,7 +458,181 @@ describe('instrument', () => {
...
+          )
+        })
+
+        const baseVariables = {
+          DD_KMS_API_KEY: '5678',
+          DD_MERGE_XRAY_TRACES: 'false',
+          DD_SITE: 'datadoghq.com',
+          DD_TRACE_ENABLED: 'false',
+        }
+        const compatibleTradeAndExtension = {
```
Comment (DarcyRaynerDD) [inline src/commands/lambda/__tests__/functions/instrument.test.ts:505]: I think this is a typo, `compatibleTradeAndExtension` -> `compatibleTracerAndExtension`

File: src/commands/lambda/__tests__/functions/instrument.test.ts [hunk 5]
```diff
@@ -458,7 +458,181 @@ describe('instrument', () => {
...
+            'arn:aws:lambda:us-east-1:123456789012:layer:dd-trace-dotnet:4',
+          ],
+        }
+        const oldExtensionVersion = {
+          Environment: {
+            Variables: {
+              ...baseVariables,
+              CORECLR_ENABLE_PROFILING: '1',
+              CORECLR_PROFILER: '{846F5F1C-F9AE-4B07-969E-05C26BC060D8}',
+              CORECLR_PROFILER_PATH: '/opt/datadog/Datadog.Trace.ClrProfiler.Native.so',
```
Comment (DarcyRaynerDD) [inline src/commands/lambda/__tests__/functions/instrument.test.ts:521]: @IvanTopolcic  mentioned this CORECLR_PROFILER_PATH path was changing in the next release of the dotnet-tracer. I think it will include something like `linux-x86' in the path, or `linux-arm` for ARM. Is that right Ivan?

File: src/commands/lambda/__tests__/functions/instrument.test.ts [hunk 6]
```diff
@@ -458,7 +458,181 @@ describe('instrument', () => {
...
+          async ({extensionVersion, traceVersion, outputResult}) => {
+            const curSettings = {...settings, extensionVersion, layerVersion: traceVersion}
+            process.env[CI_KMS_API_KEY_ENV_VAR] = '5678'
+            const updateRequest = await calculateUpdateRequest(dotNetConfig, curSettings, region, dotnetRuntime)
+            expect(updateRequest).toEqual(outputResult)
+          }
+        )
+      })
+
+      describe('test for runtime ${javaRuntime}', () => {
```
Comment (DarcyRaynerDD) [inline src/commands/lambda/__tests__/functions/instrument.test.ts:561]:
```suggestion
      describe(`test for runtime ${javaRuntime}`, () => {
```

File: src/commands/lambda/__tests__/functions/instrument.test.ts [hunk 7]
```diff
@@ -458,7 +458,181 @@ describe('instrument', () => {
...
+            const curSettings = {...settings, extensionVersion, layerVersion: traceVersion}
+            process.env[CI_KMS_API_KEY_ENV_VAR] = '5678'
+            const updateRequest = await calculateUpdateRequest(javaConfig, curSettings, region, javaRuntime)
+            expect(updateRequest).toEqual(outputResult)
+          }
+        )
+      })
+    })
   })
+
```
Comment (DarcyRaynerDD) [inline src/commands/lambda/__tests__/functions/instrument.test.ts:635]: This file is now > 1000 lines. Might make sense to split the tests up into a few files for readability.

File: src/commands/lambda/__tests__/functions/versionChecker.test.ts [hunk 8]
```diff
@@ -0,0 +1,41 @@
+import {RuntimeType} from '../../constants'
+import {
+  isExtensionCompatibleWithTrace,
+  isExtensionSupportUniversalInstrumentation,
+} from '../../functions/versionChecker'
+
+describe('Test extension and trace version checker', () => {
+  test.each`
```
Comment (DarcyRaynerDD) [inline src/commands/lambda/__tests__/functions/versionChecker.test.ts:8]: Nice use of a matrix test.

File: src/commands/lambda/functions/versionChecker.ts [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+    case RuntimeType.JAVA:
+      return extensionVersion >= UNIVERSAL_INSTRUMENT_JAVA_EXTENSION_VERSION
+    case RuntimeType.DOTNET:
+      return extensionVersion >= UNIVERSAL_INSTRUMENT_DOTNET_EXTENSION_VERSION
+    default:
+      return true
+  }
+}
+
+export const isExtensionCompatibleWithTrace = (runtimeType: RuntimeType, traceVersion?: number): boolean => {
```
Comment (DarcyRaynerDD) [inline src/commands/lambda/functions/versionChecker.ts:25]: I would suggest `isTracerCompatibleWithExtension`, since is accepting a tracer version as an argument.

File: src/commands/lambda/functions/versionChecker.ts [hunk 10]
```diff
@@ -0,0 +1,38 @@
+import {RuntimeType} from '../constants'
+
+const UNIVERSAL_INSTRUMENT_JAVA_EXTENSION_VERSION = 24
```
Comment (DarcyRaynerDD) [inline src/commands/lambda/functions/versionChecker.ts:3]: `UNIVERSAL_INSTRUMENT_JAVA_EXTENSION_VERSION` -> `UNIVERSAL_INSTRUMENTATION_JAVA_EXTENSION_VERSION`

**Review by DarcyRaynerDD — APPROVED** [review]
Overall looks good. Had some general nits about naming, and one question about the correct paths for the .NET tracer

**Comment by nine5two7** [comment]
> Overall looks good. Had some general nits about naming, and one question about the correct paths for the .NET tracer

Thanks @DarcyRaynerDD, I will create a new revision later (after Ivan confirms the path change for dotnet) based on you comments.

File: src/commands/lambda/__tests__/functions/instrument.test.ts (hunk 7 above)
Comment (nine5two7) [reply src/commands/lambda/__tests__/functions/instrument.test.ts:635]: Totally agree. Some kind of refactoring is also necessary now.

File: src/commands/lambda/__tests__/functions/instrument.test.ts (hunk 4 above)
Comment (nine5two7) [reply src/commands/lambda/__tests__/functions/instrument.test.ts:505]: Good catch

File: src/commands/lambda/__tests__/functions/instrument.test.ts (hunk 5 above)
Comment (nine5two7) [reply src/commands/lambda/__tests__/functions/instrument.test.ts:518]: For newer extension versions, it was handled in datadog_wrapper.
For old extension versions, it was left as it was.

**Comment by nine5two7** [comment]
Pass all unit tests and integration tests
2) Manual verified in both interactive and interactive mode
3) Manually tested for the corner case when tracer is not compatible with extension

File: src/commands/lambda/functions/instrument.ts [hunk 11]
```diff
@@ -267,47 +257,72 @@ export const calculateUpdateRequest = async (
...
       extensionVersion = await findLatestLayerVersion(EXTENSION_LAYER_KEY as LayerKey, region)
     }
     fullExtensionLayerARN = `${lambdaExtensionLayerArn}:${extensionVersion}`
   }
-
   layerARNs = addLayerArn(fullExtensionLayerARN, DD_LAMBDA_EXTENSION_LAYER_NAME, layerARNs)
 
+  // Special handling for .NET and Java to support universal instrumentation
+  if (runtimeType === RuntimeType.DOTNET || runtimeType === RuntimeType.JAVA) {
+    if (layerOrTraceVersion && isExtensionSupportUniversalInstrumentation(runtimeType, extensionVersion)) {
```
Comment (DarcyRaynerDD) [inline src/commands/lambda/functions/instrument.ts:292]: How about `extensionSupportsUniversalInstrumentation` or 
`isExtensionCompatibleWithUniversalInstrumentation`. 

isExtensionSupportUniversalInstrumentation sounds grammatically incorrect to me.

File: src/commands/lambda/functions/instrument.ts [hunk 12]
```diff
@@ -267,47 +257,72 @@ export const calculateUpdateRequest = async (
...
+        )
+      }
+    } else if (runtimeType === RuntimeType.DOTNET) {
+      // If it is an old extension version or the trace version is null, leave it is as the old workflow
+      if (
+        !isExtensionSupportUniversalInstrumentation(runtimeType, extensionVersion) &&
+        config.Architectures?.includes(ARM64_ARCHITECTURE)
+      ) {
+        throw new Error(
+          'Instrumenting arm64 architecture is not supported for the given dd-extension version. Please choose the latest dd-extension version or use x86_64 architecture.'
```
Comment (DarcyRaynerDD) [inline src/commands/lambda/functions/instrument.ts:310]: Is this correct? I believe we released the ARM .NET versions of the layer. @IvanTopolcic  since Storms is out

**Review by DarcyRaynerDD — CHANGES_REQUESTED** [review]
Looks good, had one question and a nit.

File: src/commands/lambda/functions/instrument.ts (hunk 12 above)
Comment (nine5two7) [reply src/commands/lambda/functions/instrument.ts:310]: In the origin code, it throws an error for the combination of .NET and ARM64. My understanding is that our old tracer&extension versions do not support .NET running on ARM64. However, the latest version of extension and tracer supports.  My implementation was based on this understanding.

Comment (DarcyRaynerDD) [reply src/commands/lambda/functions/instrument.ts:310]: Ahh, that makes sense

**Review by DarcyRaynerDD — APPROVED** [review]
